torch/csrc/distributed/c10d/init.cpp

```diff
@@ -38,6 +38,42 @@
 
 #include <torch/custom_class.h>
 
+// Wrapper to ensure GIL is released before destructing ProcessGroupGloo
+template <typename T>
+class IntrusivePtrNoGilDestructor {
+  c10::intrusive_ptr<T> impl_;
+public:
+  IntrusivePtrNoGilDestructor() = default;
+  IntrusivePtrNoGilDestructor(const IntrusivePtrNoGilDestructor&) = default;
+  IntrusivePtrNoGilDestructor(IntrusivePtrNoGilDestructor&&) = default;
+  IntrusivePtrNoGilDestructor& operator=(const IntrusivePtrNoGilDestructor&) = default;
+  IntrusivePtrNoGilDestructor& operator=(IntrusivePtrNoGilDestructor&&) = default;
+  IntrusivePtrNoGilDestructor(c10::intrusive_ptr<T> impl) : impl_(std::move(impl)) {}
+  // This ctor is very important; see
+  // https://github.com/pybind/pybind11/issues/2957
+  IntrusivePtrNoGilDestructor(T* impl) : impl_(c10::intrusive_ptr<T>::reclaim(impl)) {
+    c10::raw::intrusive_ptr::incref(impl);
```

Comment:
@wanchaol I'm tagging you specifically for this line, this replicates the logic you wrote in https://github.com/pytorch/pytorch/pull/47586/files#diff-a8f056e4406fefdc132b5ba55fcebb64757224f649a1acd86b1861fa4fb3a5bcR246 because your implementation is not public. Should we be adding another static method for this specific use case? Note that I *cannot* use `unsafe_reclaim_from_nonowning` as this method requires refcount to be nonzero, which is not the case here (the raw pointer constructor is being used in the classic unique pointer manner)

Reply:
Yeah it looks reasonable to have a general static function to be used by both. For this line specifically, it seems not increasing the weak_count, if making this a general function, i guess weak_count also need to be increased?

Reply:
incref increases weakcount, not to worry :)